src/sparse/merkle_tree.rs

```diff
@@ -0,0 +1,385 @@
+                    } else {
+                        continue;
+                    }
+                } else {
+                    side_nodes[i - offset_side_nodes].clone()
+                }
+            };
+
+            let requested_leaf_key = requested_leaf_node.leaf_key();
+            if requested_leaf_key
+                .get_bit_at_index_from_msb(self.depth() - 1 - i)
+                .unwrap()
+                == 1
+            {
+                current_node = Node::create_node(&side_node, &current_node);
+            } else {
+                current_node = Node::create_node(&current_node, &side_node);
+            }
+            self.storage
+                .insert(&current_node.hash(), current_node.as_buffer())?;
+        }
+
+        self.root_node = current_node;
+
+        Ok(())
+    }
+
+    fn delete_with_path_set(
+        &'a mut self,
+        requested_leaf_node: &Node,
+        path_nodes: &[Node],
+        side_nodes: &[Node],
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        for node in path_nodes {
+            self.storage.remove(&node.hash())?;
+        }
+
+        let mut non_placeholder_reached = false;
+        let mut current_node = Node::create_placeholder();
+        let n = side_nodes.len();
+        for i in 0..n {
+            let side_node = &side_nodes[i];
+            if current_node.is_placeholder() {
+                if side_node.is_leaf() {
+                    current_node = side_node.clone();
+                    continue;
+                } else {
+                    non_placeholder_reached = true;
+                }
+            }
+
+            if !non_placeholder_reached && side_node.is_placeholder() {
+                continue;
+            } else if !non_placeholder_reached {
+                non_placeholder_reached = true;
+            }
+
+            let requested_leaf_key = requested_leaf_node.leaf_key();
+            if requested_leaf_key
+                .get_bit_at_index_from_msb(self.depth() - 1 - i)
+                .unwrap()
+                == 1
+            {
+                current_node = Node::create_node(&side_node, &current_node);
+            } else {
+                current_node = Node::create_node(&current_node, &side_node);
+            }
+            self.storage
+                .insert(&current_node.hash(), current_node.as_buffer())?;
+        }
+
+        self.root_node = current_node;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::common::{Bytes32, StorageError, StorageMap};
+    use crate::sparse::{Buffer, MerkleTree};
+    use hex;
+
+    #[test]
+    fn test_root_returns_empty_sum_with_no_leaves() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let tree = MerkleTree::<StorageError>::new(&mut storage).unwrap();
+        let root = tree.root();
+        let expected_root = "0000000000000000000000000000000000000000000000000000000000000000";
+        assert_eq!(hex::encode(root), expected_root);
```

Comment:
Based on the celestia [specs](https://github.com/celestiaorg/celestia-specs/blob/master/src/specs/data_structures.md#sparse-merkle-tree), shouldn't the root of an empty tree be this? 

> The base case (an empty tree) is defined as the [hash](https://github.com/celestiaorg/celestia-specs/blob/master/src/specs/data_structures.md#hashing) of the empty string:
```
node.v = 0xe3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

Reply:
It looks like the default root is set to this here: https://github.com/celestiaorg/smt/blob/d95f5f42e254b8e79e09de7da97e08795a003a94/smt.go#L37

Reply:
It's actually a bit misleading: the spec and the code actually point to different results. The default root, as set by
```go
smt.SetRoot(smt.th.placeholder())
```
is the placeholder. A placeholder, in both the Go and Rust implementations, is a 32 byte empty array (i.e. `32 \0`s). As you correctly point out, this is contradictory to the spec. In the case of this discrepancy, I chose to go with the code output.

I verified this by adding the following test code in my local Celestia SMT repo:
```go
func TestEmptyRoot(t *testing.T) {
	smt := NewSparseMerkleTree(NewSimpleMap(), NewSimpleMap(), sha256.New())
	t.Logf("ROOT: %x", smt.Root())
}
```

This discrepancy should be fixed in Celestia (either in code or in spec), but ideally Fuel should have its own spec that governs these cases definitively.

Reply:
If we assume the correct root of an empty tree is the hash of the empty string, the `root()` function could be as simple as checking if the root is empty, and returning this correct hash, instead. However, the hash of a placeholder must be `000...`. We cannot assert that the hash of a placeholder is also the hash of the empty string. Otherwise, we accumulate hashes of the empty string.

Reply:
I created an issue in Celestia for this: https://github.com/celestiaorg/smt/issues/67. 

My goal is to write the Fuel implementation with the flexibility to handle future changes with relative ease. I agree with the idea of abstracting these things. (I had previously introduced separate getters for the "empty sum" and the "zero sum", where the empty sum is the hash of the empty string. I removed it in this PR because the empty sum wasn't actually used, but I will bring it back in an upcoming commit.)

Reply:
Ok well I'm fine with this as is if it matches the behavior of the actual celestia implementation for now. Once we know whether the specs are right or not about this we can fix both celestia and this later.